batik-util/src/main/java/org/apache/batik/util/CleanerThread.java

```diff
@@ -99,25 +102,30 @@ protected CleanerThread() {
     }
 
     public void run() {
-        while(true) {
-            try {
-                Reference ref;
-                try {
-                    ref = queue.remove();
+        try {
+            while(!shutdownLatch.await(100, TimeUnit.MILLISECONDS)) {
+                Reference<?> ref;
+                do {
+                    ref = queue.poll();
                     // System.err.println("Cleaned: " + ref);
-                } catch (InterruptedException ie) {
-                    continue;
-                }
-
-                if (ref instanceof ReferenceCleared) {
-                    ReferenceCleared rc = (ReferenceCleared)ref;
-                    rc.cleared();
-                }
-            } catch (ThreadDeath td) {
-                throw td;
-            } catch (Throwable t) {
-                t.printStackTrace();
+                    if (ref != null && ref instanceof ReferenceCleared) {
+                        ReferenceCleared rc = (ReferenceCleared)ref;
+                        rc.cleared();
+                    }
+                } while (ref != null);
+            }
+        } catch (ThreadDeath td) {
+            throw td;
+        } catch (Throwable t) {
+            t.printStackTrace();
+        } finally {
```

Comment:
You should release the hatch at the `finally` block, before setting `queue` to `null`:
```java
    shutdownLatch.countDown();
```